test/test_distributed.py

```diff
@@ -484,21 +551,150 @@ def _test_barrier_helper(self, group, group_id, rank):
+
+        self._barrier()
+
+    @unittest.skipIf(BACKEND != 'nccl',
+                     "Only Nccl backend supports allreduce multigpu")
+    @skip_if_no_cuda_distributed
+    @skip_if_single_gpu
+    def test_all_reduce_multigpu(self):
+        group, group_id, rank = self._init_global_test()
+        visible_gpus_per_rank = self._init_multigpu_helper()
+        self._test_all_reduce_multigpu_helper(
+            group, group_id, rank, visible_gpus_per_rank, dist.reduce_op.SUM,
+            2, 10, (2 + 10 * (len(group) - 1)) * len(visible_gpus_per_rank[0]))
+
+    def _test_reduce_multigpu_helper(self, group, group_id, rank,
+                                     visible_gpus_per_rank, op, master_value,
+                                     worker_value, expected_value):
+        for src in group:
+            if rank == src:
+                tensors = [_build_tensor(src + 1, master_value).cuda(device=i)
+                           for i in visible_gpus_per_rank[rank]]
+                dist.reduce_multigpu(tensors, src, op, group_id)
+                expected_tensor = _build_tensor(src + 1, expected_value)
+                self.assertEqual(tensors[0], expected_tensor)
+            else:
+                tensors = [_build_tensor(src + 1, worker_value).cuda(device=i)
+                           for i in visible_gpus_per_rank[rank]]
+                dist.reduce_multigpu(tensors, src, op, group_id)
+
+        self._barrier()
+
+    @unittest.skipIf(BACKEND != 'nccl',
+                     "Only Nccl backend supports reduce multigpu")
+    @skip_if_single_gpu
+    @skip_if_no_cuda_distributed
+    def test_reduce_multigpu(self):
+        group, group_id, rank = self._init_global_test()
+        visible_gpus_per_rank = self._init_multigpu_helper()
+        self._test_reduce_multigpu_helper(
+            group, group_id, rank, visible_gpus_per_rank, dist.reduce_op.SUM,
+            2, 10, (2 + 10 * (len(group) - 1)) * len(visible_gpus_per_rank[0]))
+
+    def _test_all_gather_multigpu_helper(self, group, group_id, rank,
+                                         visible_gpus_per_rank):
+        for dest in group:
+            tensors = [_build_tensor(dest + 1).cuda(device=i)
+                       for i in visible_gpus_per_rank[rank]]
+
+            output_r = []
+            expected_output = _build_tensor(dest + 1)
+            for r in visible_gpus_per_rank.keys():
+                for gpu in visible_gpus_per_rank[r]:
+                    output_r.append(_build_tensor(dest + 1, 0))
+
+            output_tensors = []
+            for gpu in visible_gpus_per_rank[rank]:
+                gpu_tensors = [t.cuda(device=gpu) for t in output_r]
+                output_tensors.append(gpu_tensors)
+
+            dist.all_gather_multigpu(output_tensors, tensors, group_id)
+            for tensor_list in output_tensors:
+                for t in tensor_list:
+                    self.assertEqual(t, expected_output)
```

Comment:
I don't fully understand this test, can you please add some comments?

Reply:
This is to test all_gather_multigpu() function. https://github.com/pytorch/pytorch/blob/master/torch/distributed/__init__.py#L408
In the context of multigpu test setup described in the comment above.

Reply:
I undestand how the tests are set up. I don't understand what e.g. `output_r` (in contrast to `output_tensors`) is.